Fix FSDP device_id when CPU offloading

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82892

See https://github.com/pytorch/pytorch/issues/82891 for full context. 

When we init FSDP with device_id + CPU offload, we could potentially hit a crash when an outer FSDP unit does not manage any params. What was happening is that it would end up getting a flat param of a child FSDP module, check the device of this, see it is CPU, and throw an error.

The fix is to avoid this check if we hit a flat param. Also fixes up the documentation of the function.

rohan-varma committed:
Fix

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_misc.py:
```diff
@@ -118,6 +118,35 @@ def test_device_id_auto_wrap(self):
+
+        cpu_device = torch.device("cpu")
+        for fsdp_unit in FSDP.fsdp_modules(fsdp):
+            fsdp_param = next(fsdp_unit.parameters())
```
nit: This is subtly relies on the order of `.parameters()`. Maybe we should do `fsdp_param = fsdp_unit.params[0]`?

rohan-varma replied:
Curious, what does `.params` give if the current FSDP unit manages no parameters?

awgu replied:
it is just an empty list then. It is not safe to access `.params[0]` in general, but I recommended it since it was fine for your particular model.

awgu reviewed (approved):
Nice catch!

rohan-varma committed:
Update on "Fix FSDP device_id when CPU offloading"


See https://github.com/pytorch/pytorch/issues/82891 for full context. 

When we init FSDP with device_id + CPU offload, we could potentially hit a crash when an outer FSDP unit does not manage any params. What was happening is that it would end up getting a flat param of a child FSDP module, check the device of this, see it is CPU, and throw an error.

The fix is to avoid this check if we hit a flat param. Also fixes up the documentation of the function.

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix FSDP device_id when CPU offloading"


See https://github.com/pytorch/pytorch/issues/82891 for full context. 

When we init FSDP with device_id + CPU offload, we could potentially hit a crash when an outer FSDP unit does not manage any params. What was happening is that it would end up getting a flat param of a child FSDP module, check the device of this, see it is CPU, and throw an error.

The fix is to avoid this check if we hit a flat param. Also fixes up the documentation of the function.

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge
